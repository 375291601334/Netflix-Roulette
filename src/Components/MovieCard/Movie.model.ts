export type Movie = {
  id: number;
  title: string;
  tagline: string;
  vote_average: number;
  vote_count: number;
  release_date: string;
  poster_path: string;
  overview: string;
  budget: number;
  revenue: number;
  genres: string[];
  runtime: number;
};
