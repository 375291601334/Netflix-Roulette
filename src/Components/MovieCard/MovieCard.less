@import '../../Styles/variables.less';

.cardContainer {
  position: relative;
  box-sizing: border-box;
  width: 33.33%;
  padding: 20px 2.5%;

  .tablet({
    width: 50%;
  });

  .mobile({
    width: 100%;
    padding: 20px 0;
  });

  &:hover {
    cursor: pointer;
    filter: brightness(70%);

    .menuIcon {
      display: block;

      &:hover {
        filter: brightness(170%);
      }
    }
  }

  .menuIcon {
    display: none;
    position: absolute;
    top: 30px;
    right: 11%;
    padding: 5px 20px;
    font-size: 40px;
    background: @darkGrey;
    border-radius: 50%;
    box-shadow: 0 0 10px @grey;
  }

  .menu {
    display: none;
    position: absolute;
    top: 5%;
    right: 11%;
    width: 40%;
    margin: 10px;
    padding: 30px 0 10px;
    background: @darkGrey;
    border-radius: 5px;

    .option {
      padding: 10px 30px;

      &:hover {
        background: @red;
        cursor: pointer;
      }
    }

    .closeButton {
      position: absolute;
      top: 0;
      right: 5px;
      background: none;
      cursor: pointer;

      &:hover .closeIcon {
        color: @red;
      }
    }

    .closeIcon {
      display: block;
      font-size: 30px;
      color: @white;
      transform: rotate(45deg);
    }

    .tablet({
      width: 70%;

      .option {
        padding: 10px 20px;
      }
    });

    .mobile({
      .option {
        padding: 20px;
        font-size: 20px;
      }
    });
  }
}

.card {
  height: 100%;
  display: flex;
  flex-direction: column;
  color: @white;

  img {
    flex: auto;
    width: 100%;
    min-height: 200px;
    margin-bottom: 20px;
    background-color: @black;
  }

  .mainInfoContainer {
    display: flex;
    justify-content: space-between;
    margin-bottom: 10px;
  }

  h4 {
    font-size: 20px;
  }

  .date {
    padding: 5px 10px;
    border: 1px solid @white;
    border-radius: 3px;
  }
}
