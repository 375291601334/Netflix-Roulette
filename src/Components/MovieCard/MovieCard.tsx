import React from 'react';
import PropTypes from 'prop-types';
import { useRouter } from 'next/router';
import { ModalWithForm } from '../ModalWithForm';
import { DeleteModal } from '../DeleteModal';
import { ModalWindowWrapper } from '../ModalWindowWrapper';
import { Movie } from './Movie.model';
import { useToggle } from '../../Utilities';
import css from './MovieCard.less';

export function MovieCard({ movie, deleteMovie, editMovie }:
  {
    movie: Movie,
    deleteMovie: (id: Movie['id']) => void,
    editMovie: (movie: Movie) => void,
  },
) {
  const router = useRouter();
  const [showEditModal, setShowEditModal] = useToggle(false);
  const [showDeleteModal, setShowDeleteModal] = useToggle(false);

  const openMenu = (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
    ((event.target as HTMLElement).nextElementSibling as HTMLElement).style.setProperty('display', 'block');
  };

  const closeMenu = (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
    ((event.currentTarget as HTMLElement).parentElement as HTMLElement).style.setProperty('display', 'none');
  };

  const openEditModal = (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
    setShowEditModal();
    closeMenu(event);
  };

  const openDeleteModal = (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
    setShowDeleteModal();
    closeMenu(event);
  };

  const onEditSubmit = (formData: Partial<Movie>) => {
    editMovie({ ...movie, ...formData });
    setShowEditModal();
  };

  const onDeleteConfirm = () => {
    deleteMovie(movie.id);
    setShowDeleteModal();
  };

  const goToMovieDetails = () => {
    router.push(`/films/${movie.id}`);
  };

  return (
    <>
      <div className={css.cardContainer}>
        <span className={css.menuIcon} onClick={openMenu}>â‹®</span>
        <div className={css.menu}>
          <button className={css.closeButton} onClick={closeMenu}>
            <span className={css.closeIcon}>+</span>
          </button>
          <div className={css.option} onClick={openEditModal}>Edit</div>
          <div className={css.option} onClick={openDeleteModal}>Delete</div>
        </div>
        <div className={css.card} onClick={goToMovieDetails}>
          <img src={movie.poster_path} alt={`${movie.title} poster`} />
          <div className={css.mainInfoContainer}>
            <h4>{movie.title}</h4>
            <span className={css.date}>
              {new Date(movie.release_date).getFullYear()}
            </span>
          </div>
          <p>{movie.genres.join(', ')}</p>
        </div>
      </div>
      {
        showDeleteModal ? (
          <ModalWindowWrapper onClose={setShowDeleteModal}>
            <DeleteModal onConfirm={onDeleteConfirm} />
          </ModalWindowWrapper>
        ) : null
      }
      {
        showEditModal ? (
          <ModalWindowWrapper onClose={setShowEditModal}>
            <ModalWithForm
              movie={movie}
              onReset={setShowEditModal}
              onSubmit={onEditSubmit}
            />
          </ModalWindowWrapper>
        ) : null
      }
    </>
  );
}

MovieCard.propTypes = {
  movie: PropTypes.exact({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    tagline: PropTypes.string,
    vote_average: PropTypes.number,
    vote_count: PropTypes.number,
    release_date: PropTypes.string,
    poster_path: PropTypes.string.isRequired,
    overview: PropTypes.string.isRequired,
    budget: PropTypes.number,
    revenue: PropTypes.number,
    genres: PropTypes.arrayOf(PropTypes.string).isRequired,
    runtime: PropTypes.number.isRequired,
  }).isRequired,
  deleteMovie: PropTypes.func.isRequired,
  editMovie: PropTypes.func.isRequired,
};
