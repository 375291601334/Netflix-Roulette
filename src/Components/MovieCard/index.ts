export { MovieCard } from './MovieCard';
export { Movie } from './Movie.model';
