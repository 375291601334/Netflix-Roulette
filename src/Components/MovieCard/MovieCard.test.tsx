import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/react';
import { MovieCard } from './MovieCard';

const deleteMovie = jest.fn();
const editMovie = jest.fn();

const movie = {
  id: 123,
  title: 'Movie title',
  tagline: '',
  vote_average: 7.3,
  vote_count: 8.5,
  release_date: '2020-10-13',
  poster_path: 'https://someImage.com',
  overview: 'Movie overview',
  budget: 4365600,
  revenue: 5466453,
  genres: ['Comedy'],
  runtime: 123,
};

jest.mock('../ModalWindowWrapper', () => ({
  ModalWindowWrapper: ({ children, onClose }) => {
    return (
      <div className="modal">
        <h2>Modal Window</h2>
        <button onClick={onClose}>Close</button>
        {children}
      </div>
    );
  },
}));

jest.mock('../ModalWithForm', () => ({
  ModalWithForm: ({ onReset, onSubmit }) => {
    return (
      <div>
        <h2>Edit Movie Form</h2>
        <button onClick={onSubmit}>SUBMIT</button>
        <button onClick={onReset}>RESET</button>
      </div>
    );
  },
}));

describe('MovieCard', () => {
  test('render MovieCard', () => {
    const { baseElement } = render(
      <BrowserRouter>
        <MovieCard editMovie={editMovie} deleteMovie={deleteMovie} movie={movie} />
      </BrowserRouter>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  test('open and close movie menu', () => {
    const { getByText, container } = render(
      <BrowserRouter>
        <MovieCard editMovie={editMovie} deleteMovie={deleteMovie} movie={movie} />
      </BrowserRouter>,
    );

    userEvent.click(getByText('⋮'));
    expect(getByText('Edit')).toBeTruthy();
    expect(getByText('Delete')).toBeTruthy();

    userEvent.click(getByText('+'));
    expect(
      container.getElementsByClassName('menu')[0].getAttribute('style'),
    ).toBe('display: none;');
  });

  test('open and submit Edit Movie Form modal window', () => {
    const { getByText, container } = render(
      <BrowserRouter>
        <MovieCard editMovie={editMovie} deleteMovie={deleteMovie} movie={movie} />
      </BrowserRouter>,
    );

    userEvent.click(getByText('⋮'));
    userEvent.click(getByText('Edit'));
    expect(getByText('Edit Movie Form')).toBeTruthy();

    userEvent.click(getByText('SUBMIT'));
    expect(container.getElementsByClassName('modal')).toHaveLength(0);
  });

  test('open and submit Delete Movie modal window', () => {
    const { getByText, container } = render(
      <BrowserRouter>
        <MovieCard editMovie={editMovie} deleteMovie={deleteMovie} movie={movie} />
      </BrowserRouter>,
    );

    userEvent.click(getByText('⋮'));
    userEvent.click(getByText('Delete'));
    expect(getByText('DELETE MOVIE')).toBeTruthy();

    userEvent.click(getByText('CONFIRM'));
    expect(container.getElementsByClassName('modal')).toHaveLength(0);
  });
});
