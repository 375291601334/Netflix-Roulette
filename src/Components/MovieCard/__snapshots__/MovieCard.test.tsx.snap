// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard render MovieCard 1`] = `
<body>
  <div>
    <div
      class="cardContainer"
    >
      <span
        class="menuIcon"
      >
        â‹®
      </span>
      <div
        class="menu"
      >
        <button
          class="closeButton"
        >
          <span
            class="closeIcon"
          >
            +
          </span>
        </button>
        <div
          class="option"
        >
          Edit
        </div>
        <div
          class="option"
        >
          Delete
        </div>
      </div>
      <a
        class="card"
        href="/films/123"
      >
        <img
          alt="Movie title poster"
          src="https://someImage.com"
        />
        <div
          class="mainInfoContainer"
        >
          <h4>
            Movie title
          </h4>
          <span
            class="date"
          >
            2020
          </span>
        </div>
        <p>
          Comedy
        </p>
      </a>
    </div>
  </div>
</body>
`;
